proof-systems/src/darlin/dlog_acc_marlin/mod.rs

```diff
@@ -0,0 +1,936 @@
+use crate::darlin::accumulators::dlog::{DLogItem, DualDLogItem};
+use crate::darlin::dlog_acc_marlin::data_structures::{
+    DLogAccMarlinProof, DLogAccMarlinProverKey, DLogAccMarlinVerifierKey, PC,
+};
+use algebra::{to_bytes, Curve, DensePolynomial, Group, GroupVec, ToBytes, ToConstraintField};
+use digest::Digest;
+use marlin::{Error, Marlin, IOP};
+use poly_commit::fiat_shamir_rng::{FiatShamirRng, FiatShamirRngSeed};
+use poly_commit::ipa_pc::VerifierState;
+use poly_commit::optional_rng::OptionalRng;
+use poly_commit::{
+    evaluate_query_set, evaluate_query_set_to_vec,
+    ipa_pc::{CommitterKey as DLogProverKey, Parameters},
+    Evaluations, LabeledCommitment, LabeledPolynomial, LabeledRandomness, PCCommitterKey,
+    PolynomialCommitment, QuerySet,
+};
+use r1cs_core::ConstraintSynthesizer;
+use rand::RngCore;
+use std::marker::PhantomData;
+
+pub mod data_structures;
+
+pub struct DLogAccMarlin<G1: Curve, G2: Curve, D: Digest>(
+    #[doc(hidden)] PhantomData<G1>,
+    #[doc(hidden)] PhantomData<G2>,
+    #[doc(hidden)] PhantomData<D>,
+);
+
+impl<G1: Curve, G2: Curve, D: Digest> DLogAccMarlin<G1, G2, D>
+where
+    G1: Curve<BaseField = <G2 as Group>::ScalarField>
+        + ToConstraintField<<G2 as Group>::ScalarField>,
+    G2: Curve<BaseField = <G1 as Group>::ScalarField>
+        + ToConstraintField<<G1 as Group>::ScalarField>,
+    D: Digest + 'static,
+{
+    /// The personalization string for this protocol. Used to personalize the
+    /// Fiat-Shamir rng.
+    pub const PROTOCOL_NAME: &'static [u8] = b"DLOG-ACC-MARLIN-2021";
+
+    /// Generate the universal prover and verifier keys for the
+    /// argument system.
+    pub fn universal_setup(
```

Comment:
We should think of defining some kind of trait too for all the functions a Marlin variant should implement. Perhaps we can provide default implementation, re-use lot of stuff, and making the whole SW architecture more flexible and readable. What do you think guys ?

Reply:
Let us move this issue to our discussion of PCD related traits.

Reply:
As the discussion we had some time ago: we can generalize everything for sure but it's not really worth it at the moment and will lead to other time spent in refactoring rather than implementing core functionalities.
So let's postpone this need of generalizing Marlin/IOP, perhaps I will open an issue about that.